### PR TITLE
bio/b_print.c: recognize even 'j' format modifier.

### PR DESCRIPTION
**Pull request by dot-asm**
This takes further steps in #3060 context, up to point of dropping PRIu64...

### CODE REVIEW

**Comment by levitte** [comment]
(I'm waiting for the builds to finish before setting the ready label)

**Comment by bernd-edlinger** [comment]
I think the arguments are of type uint64_t while the BIO_printf
implementation uses LLONG which is a bn-type, and that
is not necessarily the same, for instance if I configure for
linux-generic32, which you might do when you know nothing
about the target.
Maybe j could mean the bn-LLONG whatever it is?
It would be good to use the ll format and make it mean uint64_t
that would be understood by gcc format checker, OTOH it might
be considered an ABI change.

**Comment by dot-asm** [comment]
> while the BIO_printf implementation uses LLONG which is a bn-type

Well, LLONG is not really bn-type, as it's defined in b_print.c. BUT! You are correct in sense that there is *dependency* on BN macros. Which in turn means that b_print.c indeed needs a "face lift". I'd suggest it to be handled as separate request though...

> Maybe j could mean the bn-LLONG whatever it is?

Whole point behind introduction of j is to have platform-neutral way to print specifically 64-bit values. Formally speaking ll could do, but gcc warning system doesn't like it. As for concern regarding bn-LLONG. Its presence is conditional and so is limb format modifier. This effectively means that ll is more or less independent of bn-LLONG. BTW, note that linux-generic32 is configured with BN_LLONG, so that all linuces look uniform and ll does prints 64-bit value *practically* universally. The only exclusion I can think of from the top of my head is 32-bit HP-UX for PA-RISC build with *vendor compiler*. But let's think if it needs to be taken into consideration...

**Comment by dot-asm** [comment]
@levitte, it occurred to me that we might need to take VMS into consideration when defining OSSLzu. As you pointed out yourself is VMS sizeof(size_t)==4 in all cases, i.e. even in LP64, which means that OSSLzu is better be defined as "u" and never as "lu". Of course VMS can be covered implicitly by `__STDC_VERSION__`, but the question is if it's possible to tell VMS compiler to adhere to earlier standard dialect that doesn't have to have z modifier.

I also will add couple of additional words to [already existing] CHANGES entry...

**Comment by dot-asm** [comment]
> which means that OSSLzu is better be defined as "u" and never as "lu". ... implicitly by `__STDC_VERSION__`

To clarify. I mean that *fallback* definition *for* VMS is better be "u" in case compiler expects pre-C99 dialect.

**Comment by levitte** [comment]
VMS should be fine, 'cause arguments are 64-bit entities anyway, so from a `va_arg` perspective, a passed `size_t` has been expanded to a 64-bit unsigned integer.

File: crypto/bio/b_print.c [hunk 1]
```diff
@@ -208,6 +208,7 @@ _dopr(char **sbuffer,
                 ch = *format++;
                 break;
             case 'q':
+            case 'j':
```
Comment (bernd-edlinger) [inline crypto/bio/b_print.c]: I think the function name `_dop`r is in a reserved name scope.

**Comment by levitte** [comment]
I've a concern, though...  `j` usually means that a `intmax_t` or `uintmax_t` argument is expected.  Can we know for sure that `LLONG` is large enough?  And also, should we make sure to cast arguments that are being formatted with `%jd` or `%ju`, in case `intmax_t` / `uintmax_t` is larger than `LLONG`?

Also, is `intmax_t` / `uintmax_t` defined everywhere?  Actually, that's a rethorical question, as I know one platform where it isn't...

**Comment by dot-asm** [comment]
> Can we know for sure that LLONG is large enough?

That's what @bernd-edlinger is talking about too. At least I think so. And suggestion to address it in separate "face lift" request. As for intmax_t availability. We don't have to rely on it, because BIO_print is *our* subroutine.

**Comment by levitte** [comment]
> We don't have to rely on it, because BIO_print is *our* subroutine.

Sure, but then we'd better define *very* precisely what `j` expects, and if need be, have a type available to cast to.

**Comment by bernd-edlinger** [comment]
I would not mind if j is not implemented at all, I for one have never used that before...

**Comment by dot-asm** [comment]
Pushed minor CHANGES update. Since it's not a code change lack of feedback will be treated as implicit approval.

File: CHANGES [hunk 2]
```diff
@@ -4,9 +4,9 @@
 
  Changes between 1.1.0e and 1.1.1 [xx XXX xxxx]
 
-  *) Add the z modifier parsing to BIO_printf() et al formatting string,
-     to be used for size_t and ssize_t (ossl_ssize_t).
-     [Richard Levitte]
+  *) Add the 'z' and 'j' modifiers to BIO_printf() et al formatting string,
+     'z' is to be used for [s]size_t, and 'j' - with [u]int64_t.
+     [Richard Levitte, Andt Polyakov]
```
Comment (levitte) [inline CHANGES]: I think you want to fix your name ;-)

File: crypto/bio/b_print.c (hunk 1 above)
Comment (levitte) [inline crypto/bio/b_print.c]: That should be addressed in a different PR

File: e_os.h [hunk 3]
```diff
@@ -30,29 +30,26 @@ extern "C" {
...
-# if (defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199901L)
+# if defined(_WIN64)
+#  define OSSLzu  "I64u"    /* One would expect _WIN{64|32} cases after
+                             * __STDC_VERSION__, but there are corner
+                             * cases of MinGW compilers that link with
+                             * non-compliant MSVCRT.DLL... */
+# elif defined(_WIN32)
+#  define OSSLzu  "u"
+# elif defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199901L
 #  define OSSLzu  "zu"
```
Comment (levitte) [inline e_os.h]: I have a sudden worry about VMS, hold on a minute...

Comment (levitte) [inline e_os.h]: Yup, VMS C RTL (absolutely latest patches included) doesn't yet have support for the `z` modifier.  I suggest VMS gets "lu".

Comment (dot-asm) [inline e_os.h]: Pushed additional commit...

File: e_os.h [hunk 4]
```diff
@@ -30,29 +30,30 @@ extern "C" {
...
+#  define OSSLzu  "I64u"    /* One would expect _WIN{64|32} cases after
+                             * __STDC_VERSION__, but there are corner
+                             * cases of MinGW compilers that link with
+                             * non-compliant MSVCRT.DLL... */
+# elif defined(_WIN32)
+#  define OSSLzu  "u"
+# elif defined(__VMS)
+#  define OSSLzu  "u"       /* VMS suffers from similar problem as MinGW,
+                             * i.e. C RTL falling behind compiler. Recall
+			     * that sizeof(size_t)==4 even in LP64 VMS. */
```
Comment (levitte) [inline e_os.h]: nit pick: change tabs to spaces...

Comment (dot-asm) [inline e_os.h]: Oops! Fixed. Thanks.
